pkg/helm/charts.go

```diff
@@ -0,0 +1,283 @@
+// SourceReference for the Helm chart.
+type ChartReference struct {
+	Chart     string
+	Version   string
+	SourceRef helmv2beta1.CrossNamespaceObjectReference
+}
+
+// DefaultChartGetter provides default ways to get a chart index.yaml based on
+// the URL scheme.
+var DefaultChartGetters = getter.Providers{
+	getter.Provider{
+		Schemes: []string{"http", "https"},
+		New:     getter.NewHTTPGetter,
+	},
+}
+
+type ChartPredicate func(*repo.ChartVersion) bool
+
+// Profiles is a predicate for scanning charts with the ProfileAnnotation.
+var Profiles = func(v *repo.ChartVersion) bool {
+	return hasAnnotation(v.Metadata, ProfileAnnotation)
+}
+
+// ScanCharts filters charts using the provided predicate.
+// TODO: Add caching based on the Status Artifact Revision.
+func (h *RepoManager) GetCharts(ctx context.Context, hr *sourcev1beta1.HelmRepository, pred ChartPredicate) ([]*pb.Profile, error) {
+	chartRepo, err := fetchIndexFile(hr.Status.URL)
+	if err != nil {
+		return nil, fmt.Errorf("fetching profiles from HelmRepository %s/%s: %w",
+			hr.GetName(), hr.GetNamespace(), err)
+	}
+
+	ps := make(map[string]*pb.Profile)
+
+	for name, versions := range chartRepo.Entries {
+		for _, v := range versions {
+			if pred(v) {
+				// if already added, update the versions array
+				if p, ok := ps[name]; ok {
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+				} else { // otherwise create a new profile and add to map
+					p = &pb.Profile{
+						Name:        name,
+						Home:        v.Home,
+						Sources:     v.Sources,
+						Description: v.Description,
+						Keywords:    v.Keywords,
+						Icon:        v.Icon,
+						KubeVersion: v.KubeVersion,
+					}
+					for _, m := range v.Maintainers {
+						p.Maintainers = append(p.Maintainers, &pb.Maintainer{
+							Name:  m.Name,
+							Email: m.Email,
+							Url:   m.URL,
+						})
+					}
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+					ps[name] = p
+				}
+			}
+		}
+	}
+
+	profiles := []*pb.Profile{}
+
+	for _, p := range ps {
+		sort.Strings(p.AvailableVersions)
+		profiles = append(profiles, p)
+	}
+
+	return profiles, nil
+}
+
+// GetValuesFile fetches the value file from a chart.
+func (h *RepoManager) GetValuesFile(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository, c *ChartReference, filename string) ([]byte, error) {
+	if err := h.updateCache(ctx, helmRepo); err != nil {
+		return nil, fmt.Errorf("updating cache: %w", err)
+	}
+
+	chart, err := h.loadChart(ctx, helmRepo, c)
+	if err != nil {
+		return nil, fmt.Errorf("loading %s from chart: %w", filename, err)
+	}
+
+	for _, v := range chart.Raw {
+		if v.Name == filename {
+			return v.Data, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find file: %s", filename)
+}
+
+func (h *RepoManager) updateCache(ctx context.Context, helmRepo *sourcev1beta1.HelmRepository) error {
+	entry, err := h.entryForRepository(ctx, helmRepo)
+	if err != nil {
+		return err
```

Comment:
We wrap the other error, might as well be consistent and wrap this one and the one for Download as well. :)

Reply:
its wrapped in the parent call to `updateCache`

Reply:
But so are the others. :D